src/core/lv2/Lv2Proc.cpp

```diff
@@ -0,0 +1,540 @@
+
+	AutoLilvNodes reqFeats(lilv_plugin_get_required_features(plugin));
+	LILV_FOREACH (nodes, itr, reqFeats.get())
+	{
+		issues.emplace_back(featureNotSupported,
+			lilv_node_as_string(lilv_nodes_get(reqFeats.get(), itr)));
+	}
+
+	if (printIssues && issues.size())
+	{
+		qDebug() << "Lv2 plugin"
+			<< qStringFromPluginNode(plugin, lilv_plugin_get_name)
+			<< "(URI:"
+			<< lilv_node_as_uri(lilv_plugin_get_uri(plugin))
+			<< ") can not be loaded:";
+		for (const PluginIssue& iss : issues) { qDebug() << "  - " << iss; }
+	}
+
+	return (audioChannels[inCount] > 2 || audioChannels[outCount] > 2)
+		? Plugin::Undefined
+		: (audioChannels[inCount] > 0)
+			? Plugin::Effect
+			: Plugin::Instrument;
+}
+
+
+
+
+Lv2Proc::Lv2Proc(const LilvPlugin *plugin, Model* parent) :
+	LinkedModelGroup(parent),
+	m_plugin(plugin)
+{
+	initPlugin();
+}
+
+
+
+
+Lv2Proc::~Lv2Proc() { shutdownPlugin(); }
+
+
+
+
+void Lv2Proc::dumpPorts()
+{
+	std::size_t num = 0;
+	for (const std::unique_ptr<Lv2Ports::PortBase>& port: m_ports)
+	{
+		(void)port;
+		dumpPort(num++);
+	}
+}
+
+
+
+
+void Lv2Proc::copyModelsFromCore()
+{
+	struct FloatFromModelVisitor : public ConstModelVisitor
+	{
+		const std::vector<float>* m_scalePointMap; // in
+		float m_res; // out
+		void visit(const FloatModel& m) override { m_res = m.value(); }
+		void visit(const IntModel& m) override {
+			m_res = static_cast<float>(m.value()); }
+		void visit(const BoolModel& m) override {
+			m_res = static_cast<float>(m.value()); }
+		void visit(const ComboBoxModel& m) override {
+			m_res = (*m_scalePointMap)[static_cast<std::size_t>(m.value())]; }
+	};
+
+	struct Copy : public Lv2Ports::Visitor
+	{
+		void visit(Lv2Ports::Control& ctrl) override
+		{
+			if (ctrl.m_flow == Lv2Ports::Flow::Input)
+			{
+				FloatFromModelVisitor ffm;
+				ffm.m_scalePointMap = &ctrl.m_scalePointMap;
+				ctrl.m_connectedModel->accept(ffm);
+				ctrl.m_val = ffm.m_res;
+			}
+		}
+		void visit(Lv2Ports::Cv& cv) override
+		{
+			if (cv.m_flow == Lv2Ports::Flow::Input)
+			{
+				FloatFromModelVisitor ffm;
+				ffm.m_scalePointMap = &cv.m_scalePointMap;
+				cv.m_connectedModel->accept(ffm);
+				// dirty fix, needs better interpolation
+				std::fill(cv.m_buffer.begin(), cv.m_buffer.end(), ffm.m_res);
+			}
+		}
+	} copy;
+
+	for (const std::unique_ptr<Lv2Ports::PortBase>& port : m_ports) {
+		port->accept(copy); }
+}
+
+
+
+
+void Lv2Proc::copyBuffersFromCore(const sampleFrame *buf,
+									unsigned offset, unsigned num,
+									fpp_t frames)
+{
+	inPorts().m_left->copyBuffersFromCore(buf, offset, frames);
+	if (num > 1)
+	{
+		// if the caller requests to take input from two channels, but we only
+		// have one input channel... take medium of left and right for
+		// mono input
+		// (this happens if we have two outputs and only one input)
+		if (inPorts().m_right)
+			inPorts().m_right->copyBuffersFromCore(buf, offset + 1, frames);
+		else
+			inPorts().m_left->addBuffersFromCore(buf, offset + 1, frames);
```

Comment:
According to the coding convention, you should use explicit blocks.

Reply:
Some questions and suggestions:
- Is `offset` the channel index? Although there's an explanation, the name is still unclear.
- The name `addBuffersFromCore` is confusing(it's taking the average). How about something like `averageWithBuffersFromCore`?

Reply:
`offset` is indeed the channel index. It's a bit confusing, but when a processor tries to write its data into `buf`, it must know for each sample frame into which sample it must write - i.e. into which index of the frame. `offset` tells the processor into which of the `m_channelsPerProc * m_procs.size()` samples of each frame of `buf` it must write its data for its first channel.

I'll explain that in the comments.

Reply:
all fixed